src/main/java/uk/gov/pay/ledger/agreement/service/AgreementService.java

```diff
@@ -0,0 +1,54 @@
+package uk.gov.pay.ledger.agreement.service;
+
+import com.google.inject.Inject;
+import uk.gov.pay.ledger.agreement.dao.AgreementDao;
+import uk.gov.pay.ledger.agreement.entity.AgreementEntity;
+import uk.gov.pay.ledger.agreement.model.Agreement;
+import uk.gov.pay.ledger.agreement.model.AgreementSearchResponse;
+import uk.gov.pay.ledger.agreement.resource.AgreementSearchParams;
+import uk.gov.pay.ledger.util.pagination.PaginationBuilder;
+
+import javax.ws.rs.WebApplicationException;
+import javax.ws.rs.core.Response;
+import javax.ws.rs.core.UriInfo;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+public class AgreementService {
+    private final AgreementDao agreementDao;
+
+    @Inject
+    public AgreementService(AgreementDao agreementDao) {
+        this.agreementDao = agreementDao;
+    }
+
+    public Optional<Agreement> findAgreement(String externalId) {
+        return agreementDao.findByExternalId(externalId)
+                .map(Agreement::from);
+    }
+
+    public AgreementSearchResponse searchAgreements(AgreementSearchParams searchParams, UriInfo uriInfo) {
+        var agreements = agreementDao.searchAgreements(searchParams)
+                .stream()
+                .map(Agreement::from)
+                .collect(Collectors.toList());
+        var total = agreementDao.getTotalForSearch(searchParams);
+
+        long size = searchParams.getDisplaySize();
+        if (total > 0 && searchParams.getDisplaySize() > 0) {
+            long lastPage = (total + size - 1) / size;
+            if (searchParams.getPageNumber() > lastPage || searchParams.getPageNumber() < 1) {
```

Comment:
I’m assuming this is copied from somewhere else.

Reply:
Yeah that's correct, it follows the exact same pattern as transaction search which is also used by the admin tool and public API. Maybe it's worth extracting this as a simple util as a following cleanup step